server/market/market.go

```diff
@@ -3,30 +3,251 @@
+// account package code can keep various data up-to-date, including order
+// status, history, cancellation statistics, etc.
+//
+// The Market performs the following:
 // - Receiving and validating new order data (amounts vs. lot size, check fees,
-//   utxos, sufficient market buy buffer, etc. with help from asset backends).
-// - Putting incoming orders into the current epoch queue, which must implement
-//   matcher.Booker so that the order matching engine can work with it.
-// - Possess an order book manager, which must also implement matcher.Booker.
+//   utxos, sufficient market buy buffer, etc.).
+// - Putting incoming orders into the current epoch queue.
+// - Maintain an order book, which must also implement matcher.Booker.
 // - Initiate order matching via matcher.Match(book, currentQueue)
 // - During and/or after matching:
 //     * update the book (remove orders, add new standing orders, etc.)
 //     * retire/archive the epoch queue
 //     * publish the matches (and order book changes?)
 //     * initiate swaps for each match (possibly groups of related matches)
-// - Continually update the order book based on data from the swap executors
-//   (e.g. failed swaps, partial fills?, etc.), communications hub (i.e. dropped
-//   clients), and other sources.
+// - Cycle the epochs.
 // - Recording all events with the archivist
+type Market struct {
+	running bool
+	ctx     context.Context
+	wg      sync.WaitGroup
 
-// The Market manager should not be overly involved with details of accounts and
-// authentication. Via the account package it should request account status with
-// new orders, verification of order signatures. The Market should also perform
-// various account package callbacks such as order status updates so that the
-// account package code can keep various data up-to-date, including order
-// status, history, cancellation statistics, etc.
-type Market struct{} // TODO
+	// Communications.
+	orderRouter chan *orderUpdateSignal  // incoming orders
+	orderFeeds  []chan *BookUpdateSignal // outgoing notifications
+
+	// Order handling: book, active epoch queues, order matcher, and swapper.
+	marketInfo    *types.MarketInfo
+	lotSize       uint64
+	book          *book.Book
+	epochDuration uint64
+	epochQueue    *EpochQueue
+	matcher       *matcher.Matcher
+	swapper       *swap.Swapper
+
+	// The asset backends.
+	baseAsset  *asset.Asset
+	quoteAsset *asset.Asset
+
+	// Persistent data storage.
+	storage db.DEXArchivist
+}
+
+func (m *Market) NewBookUpdateChan() <-chan *BookUpdateSignal {
+	bookUpdates := make(chan *BookUpdateSignal)
+	m.orderFeeds = append(m.orderFeeds, bookUpdates)
+	return bookUpdates
+}
+
+func (m *Market) SubmitOrder(ord order.Order) error {
+	return <-m.SubmitOrderAsync(ord)
+}
+
+func (m *Market) SubmitOrderAsync(ord order.Order) <-chan error {
+	sig := newOrderUpdateSignal(ord)
+	m.orderRouter <- sig
+	return sig.errChan
+}
```

Comment:
An alternative to the `orderRouter chan *orderUpdateSignal`. This channel `m.orderRouter` is for internal use, but another receive only `chan *orderUpdateSignal` could be provided to Market during construction for the order router to provide new orders to the market.

Reply:
This is the `AddEpoch` function under discussion at https://github.com/decred/dcrdex/pull/65/files#r339218225

I personally think the `Market` itself can be the `MarketTunnel`. The `DEX` could essentially create a `map[string]Market` before initializing the order router.

Reply:
Yeah, I see.  Will rename this.

Reply:
You could rename `AddEpoch` too, if you'd rather. I kinda like the more descriptive `SubmitOrderAsync`.

Reply:
Sure thing, `SubmitOrderAsync` it is.